themes/qgrid/source/js/menu.js

```diff
@@ -15,41 +15,243 @@ function toggleVisibility(e) {
+			menuItem.style.display = '';
+		}
+		else {
+			menuItem.style.display = 'none';
+		}
+	}
+}
+
+function highlightText(searchTarget, search) {
+	const { textContent } = searchTarget;
+	const index = textContent.toLowerCase().indexOf(search);
+	if (index >= 0 && search != '' && textContent[index - 1] != '/') {
+		let hightlightContent = textContent;
+		for (let i = 1; i <= search.length; i++) {
+			hightlightContent = textContent.substring(0, index) + '<span>' + textContent.substring(index, index + i) + '</span>' + textContent.substring(index + i);
+		}
+		searchTarget.innerHTML = hightlightContent;
+		let span = searchTarget.getElementsByTagName('span');
+		for (let searchSpan of span) {
+			searchSpan.classList.add('highlight');
+		}
+		return true;
+	}
+	else {
+		let span = searchTarget.getElementsByTagName('span');
+		for (let searchSpan of span) {
+			searchSpan.classList.remove('highlight');
+		}
+		return false;
+	}
+}
+
+function showTag(menuItem, tag, search) {
+	for (let searchTarget of tag.split(' ')) {
+		let indexTarget = -1;
+		let foundMatch = false;
+		for (let arraySearch of search.split(' ')) {
+			if (indexTarget < 0 && !foundMatch) {
+				if (arraySearch != '') {
+					foundMatch = true;
+					indexTarget = searchTarget.toLowerCase().indexOf(arraySearch);
+				}
+			}
+		}
+		const indexSearch = search.split(' ').indexOf(searchTarget.toLowerCase());
+		if (indexTarget >= 0 || indexSearch >= 0 || search == ' ') {
+			if (menuItem.getElementsByTagName('data-show-tag').length === 0) {
+				const newTag = document.createElement('data-show-tag');
+				newTag.appendChild(document.createElement('a'));
+				menuItem.insertBefore(newTag, menuItem.children[0]);
+			}
+			const dataShowTag = menuItem.getElementsByTagName('data-show-tag')[0];
+			dataShowTag.style.display = '';
+			const a = dataShowTag.getElementsByTagName('a')[0];
+			a.textContent = getTagsContent(tag, searchTarget);
+			a.setAttribute('href', menuItem.querySelector('.title').getAttribute('href'));
+			a.style.display = 'inline';
+			break;
+		}
+	}
+}
+
+function updateMenuLinks(search) {
 	const nav = document.getElementById('nav');
-	const li = nav.getElementsByTagName('li');
-
-	for (let i = 0; i < li.length; i++) {
-		const a = li[i].getElementsByTagName('a')[0];
-		if (a) {
-			if (a.text.toLowerCase().indexOf(filter) > -1) {
-				li[i].style.display = '';
-			} else {
-				li[i].style.display = 'none';
+	for (title of nav.querySelectorAll('.title')) {
+		const hash = '#search=';
+		let href = title.getAttribute('href');
+		const index = href.indexOf(hash);
+		if (index < 0) {
+			href += hash;
+		}
+		else {
+			if (search == '') {
+				href = href.slice(0, index);
+			}
+			else {
+				href = href.substr(0, index + hash.length);
 			}
 		}
+		title.setAttribute('href', href + search);
 	}
 }
 
+function getTagsContent(tag, searchTag) {
+	const searchTarget = tag.split(' ');
+	const index = searchTarget.indexOf(searchTag);
+	if (searchTarget.length > 2) {
```

Comment:
it's too complicated logic, please see string join operator and work with array I want to see one return statement here